cocoa/src/toga_cocoa/window.py

```diff
@@ -43,12 +44,69 @@ def windowShouldClose_(self, notification) -> bool:
         self.interface.on_close()
         return False
 
+    @objc_method
+    def windowWillClose_(self, notification) -> None:
+        # Setting the toolbar delegate to None doesn't disconnect
+        # the delegates and still triggers the delegate events.
+        # Hence, simply remove the toolbar instead.
+        if self.toolbar:
+            self.toolbar = None
+        # Disconnect the window delegate.
+        self.delegate = None
+
+        # Disconnecting the window delegate doesn't prevent custom
+        # methods like enteredMiniaturize_(which are performed with a
+        # delay i.e., having a delay != 0), from being triggered.
+        # Hence, check guards needs to be used in such custom methods.
+
     @objc_method
     def windowDidResize_(self, notification) -> None:
         if self.interface.content:
             # Set the window to the new size
             self.interface.content.refresh()
 
+    @objc_method
+    def windowDidMiniaturize_(self, notification) -> None:
+        self.performSelector(
+            SEL("enteredMiniaturize:"), withObject=None, afterDelay=0.1
+        )
+
+    @objc_method
+    def enteredMiniaturize_(self, sender) -> None:
+        if self.impl:  # pragma: no branch
+            if (
+                self.impl._pending_state_transition
+                and self.impl._pending_state_transition != WindowState.MINIMIZED
+            ):
+                self.impl._apply_state(WindowState.NORMAL)
+            else:
+                self.impl._pending_state_transition = None
+
+    @objc_method
+    def windowDidDeminiaturize_(self, notification) -> None:
+        self.impl._apply_state(self.impl._pending_state_transition)
+
+    @objc_method
+    def windowDidEnterFullScreen_(self, notification) -> None:
+        self.performSelector(SEL("enteredFullScreen:"), withObject=None, afterDelay=0)
```

Comment:
Moved the explanation comment to `windowDidEnterFullScreen_` delegate method.

Reply:
I have deferred only `apply_state(NORMAL)` and moved the rest to `windowDidEnterFullScreen_`. I had tried to avoid creating an extra method by using asyncio create_task, but it requires specifying a particular delay time, which is not always fixed and can be different on different systems. So, I have resorted to using `performSelector` instead.